sdk/go/common/resource/delete_behaviour.go

```diff
@@ -0,0 +1,23 @@
+package resource
+
+import pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+
+type DeleteBehaviour int32
+
+const (
+	DeleteBehaviourDelete  DeleteBehaviour = DeleteBehaviour(pulumirpc.DeleteBehaviour_DELETE)
+	DeleteBehaviourDrop                    = DeleteBehaviour(pulumirpc.DeleteBehaviour_DROP)
+	DeleteBehaviourProtect                 = DeleteBehaviour(pulumirpc.DeleteBehaviour_PROTECT)
```

Comment:
Protect here is the same as the protect flag. We could with 4.0 remove the protect bool flag and just use this. Also scope here to add a new Protect mode that doesn't even allow replacement (the current protect will allow resources to be replaced, just not fully deleted). Then again maybe we should split this behavior flag into DeleteBehaviour (i.e. what to do for a Delete) and ReplaceBehaviour (what to do for a Replace)? Like would it be valid to say delete replacements but drop the final delete? Protect against replacements but allow the final delete?

Reply:
Ah so it's exact same. Er, should we just use protect flag? I'm a bit too conservative perhaps - use what's out there already without changing.